### PR TITLE
fix: hnsw default partitions

### LINKED ISSUE #1662: Change HNSW default num_partitions
For small tables, we generally recommend 1 partition.

We could probably do something like: `num_rows*dim/(256*5million)` 

Also check that the IVF_PQ default in Python sync API is correct.

### PR DESCRIPTION
**Pull request by gagan-bhullar-tech**
PR fixes #1662

### CODE REVIEW

**Comment by gagan-bhullar-tech** [comment]
@wjones127 please review

File: rust/lancedb/src/index/vector.rs [hunk 1]
```diff
@@ -214,6 +214,11 @@ pub(crate) fn suggested_num_partitions(rows: usize) -> u32 {
     max(1, num_partitions)
 }
 
+pub(crate) fn suggested_num_partitions_for_hnsw(rows: usize, dim: u32) -> u32 {
+    let num_partitions = ((rows as f64) * (dim as f64)).div((256 * 5000000) as f64) as u32;
```
Comment (wjones127) [inline rust/lancedb/src/index/vector.rs:218]: I would keep as an integer. u64 should be enough.
```suggestion
    let num_partitions = (((rows as u64) * (dim as u64)) / (256 * 5_000_000)) as u32;
```

File: rust/lancedb/src/table.rs [hunk 2]
```diff
@@ -1457,6 +1453,12 @@ impl NativeTable {
             }?
         };
 
+        let num_partitions = if let Some(n) = index.num_partitions {
+            n
+        } else {
+            suggested_num_partitions_for_hnsw(self.count_rows(None).await?, num_sub_vectors)
+        };
```
Comment (wjones127) [inline rust/lancedb/src/table.rs:1460]: Could you do the same in `create_ivf_hnsw_sq_index`?

File: rust/lancedb/src/table.rs [hunk 3]
```diff
@@ -1457,6 +1453,12 @@ impl NativeTable {
             }?
         };
 
+        let num_partitions = if let Some(n) = index.num_partitions {
+            n
+        } else {
+            suggested_num_partitions_for_hnsw(self.count_rows(None).await?, num_sub_vectors)
```
Comment (wjones127) [inline rust/lancedb/src/table.rs:1459]: I think `dims` was meant to be original dimensions, not the compressed subvectors. Could you get that? It's the `n` inside of the `arrow_schema::DataType::FixedSizeList`

**Review by wjones127 — CHANGES_REQUESTED** [review]
Thanks for working on this. Have a few suggestions.

**Comment by gagan-bhullar-tech** [comment]
@wjones127 updated PR. Can you please review.

**Review by wjones127 — APPROVED** [review]

**Pull request merged**